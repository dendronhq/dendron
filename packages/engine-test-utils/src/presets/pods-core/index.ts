import JSON from "./json";
import MARKDOWN from "./markdown";

export const PODS_CORE = {
  JSON,
  MARKDOWN,
};

export const PODS_PRESETS = [{ name: "json", presets: JSON }];
