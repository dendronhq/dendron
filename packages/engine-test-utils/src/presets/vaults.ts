import { DVault } from "@dendronhq/common-all";

export const VAULTS = {
  MULTI_VAULT_WITH_THREE_VAULTS: () => {
    return [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      { fsPath: "vault3", name: "vaultThree" },
    ] as DVault[];
  },
};
