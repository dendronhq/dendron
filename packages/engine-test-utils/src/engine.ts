import {
  CleanDendronSiteConfig,
  CONSTANTS, DendronConfig, DEngineClient,
  DVault,
  DWorkspace,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings
} from "@dendronhq/common-all";
import {
  getDurationMilliseconds,
  tmpDir,
  vault2Path
} from "@dendronhq/common-server";
import {
  GenTestResults,
  PostSetupHookFunction,
  PreSetupHookFunction,
  RunEngineTestFunctionOpts,
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
  SetupTestFunctionV4,
  sinon,
  TestResult
} from "@dendronhq/common-test-utils";
import { LaunchEngineServerCommand } from "@dendronhq/dendron-cli";
import {
  createEngine as engineServerCreateEngine,
  DConfig,
  WorkspaceService
} from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import os from "os";
import path from "path";
import { SinonStub } from "sinon";
import { ENGINE_HOOKS } from "./presets";
import { GitTestUtils } from "./utils";

export type TestSetupWorkspaceOpts = {
  /**
   * Vaults to initialize engine with
   * Defaults to following if not set
   * [
   *    { fsPath: "vault1" },
   *    { fsPath: "vault2" },
   *    { fsPath: "vault3", name: "vaultThree" },
   *  ]
   */
  vaults?: DVault[];
  /**
   * Modify dendron config before initialization
   */
  modConfigCb?: (config: DendronConfig) => DendronConfig;
};

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<{engine: DEngineClient, port?: number}>;

/**
 * Create an {@link DendronEngine}
 */
export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return {engine: engineServerCreateEngine(opts) as DEngineClient, port: undefined};
}

export { DEngineClient, DVault, WorkspaceOpts };

/**
 * Create a server
 * @param opts
 * @returns
 */
// @ts-ignore
export async function createServer(opts: WorkspaceOpts) {
  return await new LaunchEngineServerCommand().enrichArgs({
    wsRoot: opts.wsRoot,
  });
}

/**
 * Create an {@link DendronEngineClient}
 */
export async function createEngineFromServer(
  opts: WorkspaceOpts
): Promise<any> {
  const { engine, port } = await createServer(opts);
  await engine.init();
  return {engine, port};
}

export function createSiteConfig(
  opts: Partial<CleanDendronSiteConfig> &
    Required<Pick<CleanDendronSiteConfig, "siteRootDir" | "siteHierarchies">>
): CleanDendronSiteConfig {
  let copts = {
    siteNotesDir: "docs",
    siteUrl: "https://localhost:8080",
    ...opts,
  };
  return {
    ...copts,
    siteIndex: DConfig.getSiteIndex(copts),
  };
}

/**
 *
 * @param opts.asRemote: add git repo
 * @param opts.wsRoot: override given wsRoot
 * @returns
 */
export async function setupWS(opts: {
  vaults: DVault[];
  workspaces?: DWorkspace[];
  asRemote?: boolean;
  wsRoot?: string;
}) {
  const wsRoot = opts.wsRoot || tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  ws.createConfig();
  const config = ws.config;
  let vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault, config, updateConfig: false });
      return vault;
    })
  );
  config.vaults = _.sortBy(config.vaults, "fsPath");
  if (config.site.duplicateNoteBehavior) {
    config.site.duplicateNoteBehavior.payload = (
      config.site.duplicateNoteBehavior.payload as string[]
    ).sort();
  }
  ws.setConfig(config);
  if (opts.workspaces) {
    const vaultsFromWs = await _.reduce(
      opts.workspaces,
      async (resp, ent) => {
        await resp;
        await WorkspaceService.createWorkspace({
          wsRoot: path.join(wsRoot, ent.name),
          vaults: ent.vaults,
        });
        return ws.addWorkspace({ workspace: ent });
      },
      Promise.resolve({ vaults: [] } as { vaults: DVault[] })
    );
    vaults = vaults.concat(vaultsFromWs.vaults);
  }
  if (opts.asRemote) {
    await GitTestUtils.createRepoWithReadme(wsRoot);
  }
  return { wsRoot, vaults };
}

export type RunEngineTestV5Opts = {
  preSetupHook?: SetupHookFunction;
  createEngine?: AsyncCreateEngineFunction;
  extra?: any;
  expect: any;
  workspaces?: DWorkspace[];
  setupOnly?: boolean;
  initGit?: boolean;
  initHooks?: boolean;
  addVSWorkspace?: boolean;
  /**
   * Path to preset wsRoot
   */
  wsRoot?: string;
} & TestSetupWorkspaceOpts;

export type RunEngineTestFunctionV5<T = any> = (
  opts: RunEngineTestFunctionOpts & { extra?: any; engineInitDuration: number, port?: number }
) => Promise<TestResult[] | void | T>;

export class TestPresetEntryV5 {
  public preSetupHook: PreSetupHookFunction;
  public postSetupHook: PostSetupHookFunction;
  public testFunc: RunEngineTestFunctionV4;
  public extraOpts: any;
  public setupTest?: SetupTestFunctionV4;
  public genTestResults?: GenTestResults;
  public vaults: DVault[];
  public workspaces: DWorkspace[];

  constructor(
    func: RunEngineTestFunctionV5,
    opts?: {
      preSetupHook?: PreSetupHookFunction;
      postSetupHook?: PostSetupHookFunction;
      extraOpts?: any;
      setupTest?: SetupTestFunctionV4;
      genTestResults?: GenTestResults;
      vaults?: DVault[];
    }
  ) {
    let {
      preSetupHook,
      postSetupHook,
      extraOpts,
      setupTest,
      genTestResults,
      workspaces,
    } = _.defaults(opts, {
      workspaces: [],
    });
    this.preSetupHook = preSetupHook ? preSetupHook : async () => {};
    this.postSetupHook = postSetupHook ? postSetupHook : async () => {};
    this.testFunc = _.bind(func, this);
    this.extraOpts = extraOpts;
    this.setupTest = setupTest;
    this.genTestResults = _.bind(
      genTestResults ? genTestResults : async () => [],
      this
    );
    this.workspaces = workspaces;
    this.vaults = opts?.vaults || [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      {
        name: "vaultThree",
        fsPath: "vault3",
      },
    ];
  }
}

/**
 *
 * To create empty workspace, initilizae with `vaults = []`
 * @param func
 * @param opts.vaults: By default, initiate 3 vaults {vault1, vault2, (vault3, "vaultThree")}
 * @param opts.preSetupHook: By default, initiate empty
 * @param opts.wsRoot: Override the randomly generated test directory for the wsRoot
 * @returns
 */
export async function runEngineTestV5(
  func: RunEngineTestFunctionV5,
  opts: RunEngineTestV5Opts
): Promise<any> {
  const {
    preSetupHook,
    extra,
    vaults: vaultsInit,
    createEngine,
    initGit,
    workspaces,
    addVSWorkspace,
  } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    // third vault has diff name
    vaults: [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      { fsPath: "vault3", name: "vaultThree" },
    ],
    addVSWorkspace: false,
  });
  try {
    // make sure tests don't overwrite local homedir contents
    TestEngineUtils.mockHomeDir();
    const { wsRoot, vaults } = await setupWS({
      vaults: vaultsInit,
      workspaces,
      wsRoot: opts.wsRoot,
    });
    if ((opts.initHooks, vaults)) {
      fs.ensureDirSync(path.join(wsRoot, CONSTANTS.DENDRON_HOOKS_BASE));
    }
    await preSetupHook({ wsRoot, vaults });
    const resp = await createEngine({ wsRoot, vaults });
    const engine = resp.engine;
    const start = process.hrtime();
    const initResp = await engine.init();
    if (addVSWorkspace) {
      fs.writeJSONSync(
        path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME),
        {
          folders: vaults.map((ent) => ({
            path: ent.fsPath,
            name: ent.name,
          })) as WorkspaceFolderRaw[],
          settings: {},
          extensions: {},
        } as WorkspaceSettings,
        { spaces: 4 }
      );
    }
    const engineInitDuration = getDurationMilliseconds(start);
    const testOpts = {
      wsRoot,
      vaults,
      engine,
      initResp,
      port: resp.port,
      extra,
      config: engine,
      engineInitDuration,
    };
    if (initGit) {
      await GitTestUtils.createRepoForWorkspace(wsRoot);
      await Promise.all(
        vaults.map((vault) => {
          return GitTestUtils.createRepoWithReadme(
            vault2Path({ vault, wsRoot })
          );
        })
      );
    }
    if (opts.setupOnly) {
      return testOpts;
    }
    const results = (await func(testOpts)) || [];
    await runJestHarnessV2(results, expect);
    return { opts: testOpts, resp: undefined, wsRoot };
  } finally {
    // restore sinon so other tests can keep running
    sinon.restore();
  }
}

export function testWithEngine(
  prompt: string,
  func: RunEngineTestFunctionV4,
  opts?: Omit<RunEngineTestV5Opts, "expect"> & { only?: boolean }
) {
  if (opts?.only) {
    return test.only(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  } else {
    return test(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  }
}

export class TestEngineUtils {
  static mockHomeDir(dir?: string): SinonStub {
    if (_.isUndefined(dir)) dir = tmpDir().name;
    return sinon.stub(os, "homedir").returns(dir);
  }

  static vault1(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault1" })!;
  }

  static vault2(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault2" })!;
  }

  static vault3(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault3" })!;
  }
}
