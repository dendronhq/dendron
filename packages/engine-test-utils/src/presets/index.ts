import { DVault, VaultUtils } from "@dendronhq/common-all";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { WorkspaceOpts } from "../engine";

export * from "./engine-server";
export * from "./pods-core";
export * from "./vaults";

export enum SETUP_HOOK_KEYS {
  /**
   * alpha: link(beta)
   * beta: link(alpha)
   */
  WITH_LINKS = "WITH_LINKS",
}

const createLink = (fname: string, opts: { vaultPrefix?: DVault }) => {
  let cVaultPrefix = "";
  if (opts.vaultPrefix) {
    cVaultPrefix = VaultUtils.toURIPrefix(opts.vaultPrefix) + "/";
  }
  return `[[${cVaultPrefix}${fname}]]`;
};

export async function callSetupHook(
  key: SETUP_HOOK_KEYS,
  opts: {
    workspaceType: "single" | "multi";
    withVaultPrefix?: boolean;
  } & WorkspaceOpts
) {
  const { workspaceType, vaults, wsRoot, withVaultPrefix } = opts;
  const isMultiVault = workspaceType !== "single";
  const cVaults = isMultiVault ? vaults : [vaults[0], vaults[0]];

  // WITH LINKS
  if (key === SETUP_HOOK_KEYS.WITH_LINKS) {
    const link1 = createLink("beta", {
      vaultPrefix: withVaultPrefix ? cVaults[1] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "alpha",
      wsRoot,
      vault: cVaults[0],
      body: link1,
    });
    const link2 = createLink("alpha", {
      vaultPrefix: withVaultPrefix ? cVaults[0] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "beta",
      wsRoot,
      vault: cVaults[1],
      body: link2,
    });
    if (isMultiVault) {
      await NoteTestUtilsV4.createNote({
        fname: "beta",
        wsRoot,
        vault: cVaults[0],
        body: link2,
        genRandomId: true,
      });
    }
  } else {
    throw Error("not supported key");
  }
}
