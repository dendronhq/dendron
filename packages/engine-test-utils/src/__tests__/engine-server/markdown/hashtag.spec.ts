import {
  AssertUtils,
  TestPresetEntryV4,
  NoteTestUtilsV4,
} from "@dendronhq/common-test-utils";
import {
  DendronASTDest,
  DendronASTTypes,
  HashTag,
  MDUtilsV5,
  UnistNode,
} from "@dendronhq/unified";
import _ from "lodash";
import {
  ConfigService,
  ConfigUtils,
  NoteDictsUtils,
  NoteProps,
  ProcFlavor,
  URI,
} from "@dendronhq/common-all";
import { runEngineTestV5 } from "../../../engine";
import { ENGINE_HOOKS } from "../../../presets";
import {
  checkNotInVFile,
  checkVFile,
  createProcCompileTests,
  createProcForTest,
  createProcTests,
  ProcTests,
} from "./utils";
import { TestConfigUtils } from "../../..";
import { TestUnifiedUtils } from "../../../utils";
import { getOpts, runTestCases } from "./v5/utils";

const { getDescendantNode } = TestUnifiedUtils;

function proc() {
  return MDUtilsV5.procRemarkParseNoData({}, { dest: DendronASTDest.HTML });
}

function runAllTests(opts: { name: string; testCases: ProcTests[] }) {
  const { name, testCases } = opts;
  describe(name, () => {
    test.each(
      testCases.map((ent) => [`${ent.dest}: ${ent.name}`, ent.testCase])
      // @ts-ignore
    )("%p", async (_key, testCase: TestPresetEntryV4) => {
      await runEngineTestV5(testCase.testFunc, {
        expect,
        preSetupHook: testCase.preSetupHook,
      });
    });
  });
}

function getHashtag(node: UnistNode): HashTag {
  return getDescendantNode<HashTag>(expect, node, 0, 0);
}

describe("hashtag", () => {
  describe("parse", () => {
    test("parses basic hashtag", () => {
      const resp = proc().parse("#my-hash-tag");
      expect(getHashtag(resp).type).toEqual(DendronASTTypes.HASHTAG);
      expect(getHashtag(resp).fname).toEqual("tags.my-hash-tag");
    });

    test("parses hashtag with hierarchy", () => {
      const resp = proc().parse("#my.hash.tag0");
      expect(getHashtag(resp).type).toEqual(DendronASTTypes.HASHTAG);
      expect(getHashtag(resp).fname).toEqual("tags.my.hash.tag0");
    });

    test("doesn't parse hashtag inside inline code block", () => {
      const resp = proc().parse("`#my-hash-tag`");
      expect(getHashtag(resp).type).toEqual("inlineCode");
    });

    test("doesn't parse hash inside a link", () => {
      const resp = proc().parse("[[#my-hash-tag]]");
      expect(getHashtag(resp).type).toEqual(DendronASTTypes.WIKI_LINK);
    });

    test("parses a hashtag in the middle of a paragraph", () => {
      const resp = proc().parse("Lorem ipsum #my-hash-tag dolor amet.");
      expect(getDescendantNode(expect, resp, 0, 1).type).toEqual(
        DendronASTTypes.HASHTAG
      );
      // @ts-ignore
      expect(getDescendantNode(expect, resp, 0, 1).value).toEqual(
        "#my-hash-tag"
      );
    });

    test("doesn't parse hashtag starting with number", () => {
      const resp = proc().parse("#123-hash-tag");
      expect(getDescendantNode(expect, resp, 0, 0).type).toEqual(
        DendronASTTypes.TEXT
      );
    });

    test("doesn't parse trailing punctuation except period (.)", () => {
      const resp1 = proc().parse(
        "Dolorem vero sed sapiente #dolores. Et quam id maxime et ratione."
      );
      expect(getDescendantNode(expect, resp1, 0, 1).type).toEqual(
        DendronASTTypes.HASHTAG
      );
      // @ts-ignore
      expect(getDescendantNode(expect, resp1, 0, 1).value).toEqual("#dolores.");

      const resp2 = proc().parse(
        "Dolorem vero sed sapiente #dolores, et quam id maxime et ratione."
      );
      expect(getDescendantNode(expect, resp2, 0, 1).type).toEqual(
        DendronASTTypes.HASHTAG
      );
      // @ts-ignore
      expect(getDescendantNode(expect, resp2, 0, 1).value).toEqual("#dolores");

      const resp3 = proc().parse(
        "Dolorem vero sed (sapiente #dolores) et quam id maxime et ratione."
      );
      expect(getDescendantNode(expect, resp3, 0, 1).type).toEqual(
        DendronASTTypes.HASHTAG
      );
      // @ts-ignore
      expect(getDescendantNode(expect, resp3, 0, 1).value).toEqual("#dolores");
    });

    test("doesn't parse trailing unicode punctuation", () => {
      const resp1 = proc().parse("彼女に「 #よろしく」言って下さい。");
      expect(getDescendantNode(expect, resp1, 0, 1).type).toEqual(
        DendronASTTypes.HASHTAG
      );
      // @ts-ignore
      expect(getDescendantNode(expect, resp1, 0, 1).value).toEqual("#よろしく");
    });

    test("doesn't parse when it's part of a sentence", () => {
      const resp1 = proc().parse("no#tag");
      expect(getDescendantNode(expect, resp1, 0, 0).type).not.toEqual(
        DendronASTTypes.HASHTAG
      );
      // @ts-ignore
      expect(getDescendantNode(expect, resp1, 0, 0).value).toEqual("no#tag");
    });
  });

  describe("rendering", () => {
    const hashtag = "#my-hash.tag0";

    describe("WHEN hashtag with prefix", () => {
      runTestCases(
        createProcCompileTests({
          name: "THEN hashtag has prefix",
          fname: "one",
          setup: async (opts) => {
            const { proc } = getOpts(opts);

            const noteCacheForRenderDict = NoteDictsUtils.createNoteDicts([
              await NoteTestUtilsV4.createNote({
                fname: "tags.my-hash.tag0",
                vault: opts.vaults[0],
                wsRoot: opts.wsRoot,
              }),
            ]);

            MDUtilsV5.setProcData(proc, { noteCacheForRenderDict });
            const resp = await proc.process(hashtag);
            return { resp, proc };
          },
          verify: {
            [DendronASTDest.HTML]: {
              [ProcFlavor.PUBLISHING]: async ({ extra }) => {
                const { resp } = extra;
                await checkVFile(resp, `href="/foo/notes/tags.my-hash.tag0"`);
              },
            },
          },
          preSetupHook: async (opts: any) => {
            const { wsRoot, vaults } = opts;
            const vault = vaults[0];
            await ENGINE_HOOKS.setupBasic(opts);
            await NoteTestUtilsV4.createNote({
              fname: "one",
              vault,
              wsRoot,
              body: hashtag,
            });
            await NoteTestUtilsV4.createNote({
              fname: "tags.my-hash.tag0",
              vault,
              wsRoot,
            });
            await TestConfigUtils.withConfig(
              (config) => {
                ConfigUtils.setPublishProp(config, "assetsPrefix", "/foo");
                return config;
              },
              { wsRoot }
            );
          },
        })
      );
    });

    const SIMPLE = createProcTests({
      name: "simple",
      setupFunc: async ({ engine, vaults, extra, wsRoot }) => {
        const config = (
          await ConfigService.instance().readConfig(URI.file(wsRoot))
        )._unsafeUnwrap();
        const proc2 = await createProcForTest({
          engine,
          dest: extra.dest,
          vault: vaults[0],
          config,
        });
        const resp = await proc2.process(hashtag);
        return { resp };
      },
      verifyFuncDict: {
        [DendronASTDest.MD_DENDRON]: async ({ extra }) => {
          const { resp } = extra;
          return [
            {
              actual: await AssertUtils.assertInString({
                body: resp.toString(),
                match: [hashtag],
              }),
              expected: true,
            },
          ];
        },
        [DendronASTDest.MD_REGULAR]: async ({ extra }) => {
          const { resp } = extra;
          return [
            {
              actual: await AssertUtils.assertInString({
                body: resp.toString(),
                match: [`[${hashtag}](tags.my-hash.tag0)`],
              }),
              expected: true,
            },
          ];
        },
        [DendronASTDest.HTML]: async ({ extra }) => {
          const { resp } = extra;
          await checkVFile(
            resp,
            '<a class="color-tag" style="--tag-color: #c95efb;" href="tags.my-hash.tag0">#my-hash.tag0</a>'
          );
        },
      },
      preSetupHook: ENGINE_HOOKS.setupBasic,
    });

    // TODO: Re-enable color test after color following ancestors in tags function is added back.
    describe.skip("colors", () => {
      test("with color", async () => {
        let note: NoteProps;
        await runEngineTestV5(
          async ({ engine, wsRoot }) => {
            const config = (
              await ConfigService.instance().readConfig(URI.file(wsRoot))
            )._unsafeUnwrap();
            const proc = await createProcForTest({
              engine,
              dest: DendronASTDest.HTML,
              vault: note.vault,
              config,
            });
            const resp = await proc.process(`#color`);
            await checkVFile(
              resp,
              '<a class="color-tag" style="--tag-color: #FF0033;" href="tags.color">#color</a>'
            );
          },
          {
            expect,
            preSetupHook: async ({ wsRoot, vaults }) => {
              note = await NoteTestUtilsV4.createNote({
                fname: "tags.color",
                wsRoot,
                vault: vaults[0],
                props: { color: "#FF0033" },
              });
            },
          }
        );
      });

      test("when enableRandomlyGeneratedColors is false, only uses explicit colors", async () => {
        let note: NoteProps;
        await runEngineTestV5(
          async ({ engine, wsRoot }) => {
            const config = (
              await ConfigService.instance().readConfig(URI.file(wsRoot))
            )._unsafeUnwrap();
            const proc = await createProcForTest({
              engine,
              dest: DendronASTDest.HTML,
              vault: note.vault,
              config,
            });
            const resp = await proc.process(`#color #uncolored`);
            await checkVFile(
              resp,
              '<a class="color-tag" style="--tag-color: #FF0033;" href="tags.color">#color</a>',
              '<a href="tags.uncolored">#uncolored</a>'
            );
          },
          {
            expect,
            preSetupHook: async ({ wsRoot, vaults }) => {
              note = await NoteTestUtilsV4.createNote({
                fname: "tags.color",
                wsRoot,
                vault: vaults[0],
                props: { color: "#FF0033" },
              });
              note = await NoteTestUtilsV4.createNote({
                fname: "tags.uncolored",
                wsRoot,
                vault: vaults[0],
              });
              await TestConfigUtils.withConfig(
                (config) => {
                  ConfigUtils.setPublishProp(
                    config,
                    "enableRandomlyColoredTags",
                    false
                  );
                  return config;
                },
                { wsRoot }
              );
            },
          }
        );
      });

      test("with color cascading from parent, self missing", async () => {
        let note: NoteProps;
        await runEngineTestV5(
          async ({ engine, wsRoot }) => {
            const config = (
              await ConfigService.instance().readConfig(URI.file(wsRoot))
            )._unsafeUnwrap();
            const proc = await createProcForTest({
              engine,
              dest: DendronASTDest.HTML,
              vault: note.vault,
              config,
            });
            const resp = await proc.process(`#parent.color`);
            await checkVFile(
              resp,
              '<a class="color-tag" style="--tag-color: #FF0033;" href="tags.parent.color">#parent.color</a>'
            );
          },
          {
            expect,
            preSetupHook: async ({ wsRoot, vaults }) => {
              note = await NoteTestUtilsV4.createNote({
                fname: "tags.parent",
                wsRoot,
                vault: vaults[0],
                props: { color: "#FF0033" },
              });
            },
          }
        );
      });

      test("with color cascading from parent, self exists", async () => {
        let note: NoteProps;
        await runEngineTestV5(
          async ({ engine, wsRoot }) => {
            const config = (
              await ConfigService.instance().readConfig(URI.file(wsRoot))
            )._unsafeUnwrap();
            const proc = await createProcForTest({
              engine,
              dest: DendronASTDest.HTML,
              vault: note.vault,
              config,
            });
            const resp = await proc.process(`#parent.color`);
            await checkVFile(
              resp,
              '<a class="color-tag" style="--tag-color: #FF0033;" href="tags.parent.color">#parent.color</a>'
            );
          },
          {
            expect,
            preSetupHook: async ({ wsRoot, vaults }) => {
              note = await NoteTestUtilsV4.createNote({
                fname: "tags.parent",
                wsRoot,
                vault: vaults[0],
                props: { color: "#FF0033" },
              });
              await NoteTestUtilsV4.createNote({
                fname: "tags.parent.color",
                wsRoot,
                vault: vaults[0],
              });
            },
          }
        );
      });

      test("overrides color cascading from parent", async () => {
        let note: NoteProps;
        await runEngineTestV5(
          async ({ engine, wsRoot }) => {
            const config = (
              await ConfigService.instance().readConfig(URI.file(wsRoot))
            )._unsafeUnwrap();
            const proc = await createProcForTest({
              engine,
              dest: DendronASTDest.HTML,
              vault: note.vault,
              config,
            });
            const resp = await proc.process(`#parent.color`);
            await checkVFile(
              resp,
              '<a class="color-tag" style="--tag-color: #00FF11;" href="tags.parent.color">#parent.color</a>'
            );
          },
          {
            expect,
            preSetupHook: async ({ wsRoot, vaults }) => {
              note = await NoteTestUtilsV4.createNote({
                fname: "tags.parent",
                wsRoot,
                vault: vaults[0],
                props: { color: "#FF0033" },
              });
              await NoteTestUtilsV4.createNote({
                fname: "tags.parent.color",
                wsRoot,
                vault: vaults[0],
                props: { color: "#00FF11" },
              });
            },
          }
        );
      });
    });

    const INSIDE_LINK = createProcTests({
      name: "inside a link",
      setupFunc: async ({ engine, vaults, extra, wsRoot }) => {
        const config = (
          await ConfigService.instance().readConfig(URI.file(wsRoot))
        )._unsafeUnwrap();
        const proc2 = await createProcForTest({
          engine,
          dest: extra.dest,
          vault: vaults[0],
          config,
        });
        const resp = await proc2.process(
          "[#dendron](https://twitter.com/hashtag/dendron)"
        );
        return { resp };
      },
      verifyFuncDict: {
        [DendronASTDest.HTML]: async ({ extra }) => {
          const { resp } = extra;
          await checkVFile(
            resp,
            '<a href="https://twitter.com/hashtag/dendron">#dendron</a>'
          );
          await checkNotInVFile(resp, `Private`);
        },
      },
      preSetupHook: ENGINE_HOOKS.setupBasic,
    });

    describe("WHEN disabled in config", () => {
      test("THEN hashtags don't get parsed or processed", async () => {
        await runEngineTestV5(
          async ({ engine, vaults, wsRoot }) => {
            const config = (
              await ConfigService.instance().readConfig(URI.file(wsRoot))
            )._unsafeUnwrap();
            const proc = MDUtilsV5.procRehypeFull(
              {
                noteToRender: (
                  await engine.findNotesMeta({
                    fname: "root",
                    vault: vaults[0],
                  })
                )[0]!,
                vault: vaults[0],
                config,
                fname: "root",
              },
              {}
            );
            const out = await proc.process("#test");
            expect(checkVFile(out, "<p>#test</p>"));
            expect(checkNotInVFile(out, "<a", "tags.test"));
          },
          {
            expect,
            preSetupHook: async ({ wsRoot }) => {
              await TestConfigUtils.withConfig(
                (config) => {
                  ConfigUtils.setWorkspaceProp(config, "enableHashTags", false);
                  return config;
                },
                { wsRoot }
              );
            },
          }
        );
      });
    });

    const ALL_TEST_CASES = [...SIMPLE, ...INSIDE_LINK];
    runAllTests({ name: "compile", testCases: ALL_TEST_CASES });
  });
});
