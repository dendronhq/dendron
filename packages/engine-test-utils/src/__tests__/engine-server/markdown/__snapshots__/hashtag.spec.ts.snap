// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hashtag rendering WHEN disabled in config THEN hashtags don't get parsed or processed 1`] = `
VFile {
  "contents": "<h1 id=\\"root\\">Root</h1>
<p>#test</p>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`hashtag rendering WHEN disabled in config THEN hashtags don't get parsed or processed 2`] = `
VFile {
  "contents": "<h1 id=\\"root\\">Root</h1>
<p>#test</p>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`hashtag rendering WHEN hashtag with prefix "HTML: THEN hashtag has prefix: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"one\\">One<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#one\\"></a></h1>
<p><a class=\\"color-tag\\" style=\\"--tag-color: #c95efb;\\" href=\\"/foo/notes/tags.my-hash.tag0\\">#my-hash.tag0</a></p>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`hashtag rendering compile "HTML: inside a link" 1`] = `
VFile {
  "contents": "<h1 id=\\"root\\">Root</h1>
<p><a href=\\"https://twitter.com/hashtag/dendron\\">#dendron</a></p>
<hr>
<strong>Children</strong>
<ol>
<li><a href=\\"bar\\">Bar</a></li>
<li><a href=\\"foo\\">Foo</a></li>
</ol>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`hashtag rendering compile "HTML: inside a link" 2`] = `
VFile {
  "contents": "<h1 id=\\"root\\">Root</h1>
<p><a href=\\"https://twitter.com/hashtag/dendron\\">#dendron</a></p>
<hr>
<strong>Children</strong>
<ol>
<li><a href=\\"bar\\">Bar</a></li>
<li><a href=\\"foo\\">Foo</a></li>
</ol>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`hashtag rendering compile "HTML: simple" 1`] = `
VFile {
  "contents": "<h1 id=\\"root\\">Root</h1>
<p><a class=\\"color-tag\\" style=\\"--tag-color: #c95efb;\\" href=\\"tags.my-hash.tag0\\">#my-hash.tag0</a></p>
<hr>
<strong>Children</strong>
<ol>
<li><a href=\\"bar\\">Bar</a></li>
<li><a href=\\"foo\\">Foo</a></li>
</ol>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;
